File: solution/224/solution.cpp

#include <string>
#include <stack>

class Solution {
public:
    void complute(std::stack<int> &nums, std::stack<char> &op) {
        if(nums.size()<2||op.size()<1)return;
        int a = nums.top();
        nums.pop();
        int b = nums.top();
        nums.pop();
        if(op.top()=='+')nums.push(a+b);
        else if(op.top()=='-')nums.push(b-a);
        op.pop();
    }
    int calculate(std::string s) {
        std::stack<int> nums;
        std::stack<char> op;
        static const int BEGIN = 0;
        static const int NUM = 1;
        static const int SIGN = 2;
        int state = BEGIN;
        int calculate_flag = 0;
        int number = 0;
        for(int i=0;i<s.length();i++){
            if(s[i] == ' ')continue;
            switch(state){
                case BEGIN:
                    if(s[i]>= '0'&& s[i]<= '9'){
                        state = NUM;
                    }else{
                        state = SIGN;
                    }
                    i--;
                    break;
                case NUM:
                    if(s[i]>= '0'&& s[i]<= '9'){
                        number = number*10 + s[i] - '0';
                    }else{
                        nums.push(number);
                        number = 0;
                        if(calculate_flag==1)complute(nums, op);
                        state = SIGN;
                        i--;
                    }
                    break;
                case SIGN:
                    if(s[i]=='+' || s[i]=='-'){
                        calculate_flag = 1;
                        op.push(s[i]);
                    }else if(s[i] == '('){
                        calculate_flag = 0;
                        state = NUM;
                    }else if(s[i] == ')'){
                        complute(nums, op);
                    }else if(s[i]>= '0'&& s[i]<= '9'){
                        state = NUM;
                        i--;
                    }
                    break;

            }
        }
        if(number!=0){nums.push(number);complute(nums, op);};
        if(nums.empty())return 0;
        return nums.top();
    }
};

File: solution/224/solution.py
from collections import deque
class Solution(object):
    def compute(self, nums, op):
        if(len(nums) < 2 or len(op) < 1):
            return
        b = nums.pop()
        a = nums.pop()
        if(op.pop()=='+'):nums.append(a+b)
        else:nums.append(a-b)


    def calculate(self, s):
        """
        :type s: str
        :rtype: int
        """
        BEGIN = 0
        SIGN = 1
        NUM = 2
        state = BEGIN
        number = 0
        calculate_flag = False
        nums = deque()
        op = deque()
        i = 0
        while i < len(s):
            if s[i] == ' ':
                i += 1
                continue
            if state == BEGIN:
                if s[i].isdigit():
                    state = NUM
                else:
                    state = SIGN
                i -= 1
            elif state == NUM:
                if s[i].isdigit():
                    number = number*10 + ord(s[i])-ord('0')
                else:
                    nums.append(number);
                    number = 0
                    if calculate_flag:self.compute(nums, op)
                    state = SIGN
                    i -= 1
            elif state == SIGN:
                if s[i].isdigit():
                    state = NUM
                    i -= 1
                elif s[i] == '(':
                    calculate_flag = False
                    state = NUM
                elif s[i] == '+' or s[i] == '-':
                    op.append(s[i])
                    calculate_flag = True
                elif s[i] == ')':
                    self.compute(nums, op)
            i += 1
        if number != 0:
            nums.append(number)
            self.compute(nums, op)
        if len(nums) == 0:
            return 0
        return nums[-1]


if __name__ == "__main__":
    s = Solution()
    print s.calculate("1+1")
